### PR TITLE
Feature/building names

### LINKED ISSUE #3130: Name Huts
Would be nice if citizen huts can be named

Atm it says "Citizen 1"
It could be editable to say "Beachut 1"
"Fancyhut 2" etc

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #3130

# Changes proposed in this pull request:
- Allow to name certain huts individually (use cases will come later)

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/client/gui/WindowHutNameEntry.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import com.minecolonies.coremod.colony.buildings.views.AbstractBuildingView;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Window for a hut name entry.
+ */
+public class WindowHutNameEntry extends Window implements ButtonHandler
+{
+    private static final String BUTTON_DONE                   = "done";
+    private static final String BUTTON_CANCEL                 = "cancel";
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowHutNameEntry.java]: aren't these in window constants?

File: src/main/java/com/minecolonies/coremod/client/gui/WindowHutNameEntry.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    @Override
+    public void onButtonClicked(@NotNull final Button button)
+    {
+        if (button.getID().equals(BUTTON_DONE))
+        {
+            final String name = findPaneOfTypeByID(INPUT_NAME, TextField.class).getText();
+            if (!name.isEmpty())
+            {
+                building.setCustomName(name);
+            }
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowHutNameEntry.java:64]: we should add an error message if checking fails - I would just setText("Empty names are not allowed!");

File: src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+/**
+ * Message to execute the renaiming of the townHall.
+ */
+public class HutRenameMessage extends AbstractMessage<HutRenameMessage, IMessage>
+{
+    /**
+     * The max length of the name.
+     */
+    private static final int MAX_NAME_LENGTH = 25;
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java]: this should be also errored when checking correctness of the input name in the window

File: src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+     * Object creation for the town hall rename message.
+     *
+     * @param colony Colony the rename is going to occur in.
+     * @param name   New name of the town hall.
+     */
+    public HutRenameMessage(@NotNull final IColony colony, final String name, final AbstractBuildingView b)
+    {
+        super();
+        this.colonyId = colony.getID();
+        this.name = (name.length() <= MAX_NAME_LENGTH) ? name : name.substring(0, SUBSTRING_LENGTH);
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java]: isn't substring end exclusive?

> The substring begins at the specified beginIndex and extends to the character at index endIndex - 1.

File: src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    public void messageOnServerThread(final HutRenameMessage message, final EntityPlayerMP player)
+    {
+        final Colony colony = ColonyManager.getColonyByDimension(message.colonyId, message.dimension);
+        if (colony != null && colony.getPermissions().hasPermission(player, Action.MANAGE_HUTS))
+        {
+            final AbstractBuilding b = colony.getBuildingManager().getBuildings().get(message.buildingId);
+
+            if (b != null)
+            {
+                message.name = (message.name.length() <= MAX_NAME_LENGTH) ? message.name : message.name.substring(0, SUBSTRING_LENGTH);
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java]: why do you check the length twice?

File: src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                b.setCustomBuildingName(message.name);
+                MineColonies.getNetwork().sendToAll(message);
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java]: this message does not seem to have a client acceptor method to be handled by messageOnClientThread

**Review by Nightenom — CHANGES_REQUESTED** [review]

File: src/main/java/com/minecolonies/coremod/client/gui/WindowHutNameEntry.java (hunk 1 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/client/gui/WindowHutNameEntry.java]: Will check

File: src/main/java/com/minecolonies/coremod/client/gui/WindowHutNameEntry.java (hunk 2 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/client/gui/WindowHutNameEntry.java:64]: Good idea

File: src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java (hunk 5 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java]: ugh, copy and paste error

File: src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java (hunk 6 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java]: ugh, copy and paste error

File: src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java (hunk 3 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java]: Here I agree

File: src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java (hunk 6 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java]: Uhm, it should only be handled on server side and not on client side

File: src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java (hunk 4 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java]: Good point

File: src/main/java/com/minecolonies/coremod/client/gui/WindowHutNameEntry.java (hunk 2 above)
Comment (Nightenom) [reply src/main/java/com/minecolonies/coremod/client/gui/WindowHutNameEntry.java:64]: Why the empty check disappeared?

File: src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java (hunk 6 above)
Comment (Nightenom) [reply src/main/java/com/minecolonies/coremod/network/messages/HutRenameMessage.java]: This one

File: src/main/resources/assets/minecolonies/lang/en_us.lang [hunk 7]
```diff
@@ -917,4 +917,8 @@ com.minecolonies.coremod.gui.workerHuts.Sifter=Sifter
 com.minecolonies.coremod.sifter.meshBroke=My current mesh broke and I started to work with the default mesh again!
 com.minecolonies.coremod.gui.workerHuts.buy=Buy
 com.minecolonies.coremod.gui.workerHuts.meshSettings=Mesh Settings
-com.minecolonies.coremod.sifter.toomuch=Set to the current daily max of %d.
\ No newline at end of file
+com.minecolonies.coremod.sifter.toomuch=Set to the current daily max of %d.
+com.minecolonies.coremod.gui.building.rename.title=Rename Your Building
```
Comment (Nightenom) [inline src/main/resources/assets/minecolonies/lang/en_us.lang:921]: Please include non-empty and the max length there

**Review by Nightenom — CHANGES_REQUESTED** [review]
The max length thing - just cap the input at 15 chars and mention it inside the `com.minecolonies.coremod.gui.building.rename.title`
We may allow longer ones with new assets (I checked that 15 is max now)

**Comment by Raycoms** [comment]
Removed send to all, I removed us empty because empty is valid, empty
resets to default

**Review by Nightenom — APPROVED** [review]

**Pull request merged**